// @framer/mattdesl.motiongrid/code/MotionGrid.tsx

import * as React from "react";
import { PropertyControls, ControlType } from "framer";
import * as canvasSketch from 'canvas-sketch';
import sketch, { settings } from './sketch';

const style: React.CSSProperties = {
};

export class MotionGrid extends React.Component {
  static defaultProps = {
    playing: false,
    background: 'white',
    foreground: 'black',
    stroke: true,
    gridSize: 8,
    padding: 15,
    thickness: 33,
    length: 90,
    rotation: 45,
    uMod: 20,
    vMod: 10,
    exponential: 3,
    width: 256,
    height: 256
  };

  static propertyControls: PropertyControls = {
    playing: { type: ControlType.Boolean, title: "Playing" },
    background: { type: ControlType.Color, title: "Background" },
    foreground: { type: ControlType.Color, title: "Foreground" },
    stroke: { type: ControlType.Boolean, title: "Stroke" },
    gridSize: { type: ControlType.Number, title: "Grid Size", min: 1, max: 20 },
    padding: { type: ControlType.Number, title: "Padding", min: 0, max: 40 },
    thickness: { type: ControlType.Number, title: "Thickness", min: 0, max: 100 },
    length: { type: ControlType.Number, title: "Length", min: 0, max: 100 },
    rotation: { type: ControlType.Number, title: "Rotation", min: 0, max: 180 },
    uMod: { type: ControlType.Number, title: "U Mod", min: 0, max: 100 },
    vMod: { type: ControlType.Number, title: "V Mod", min: 0, max: 100 },
    exponential: { type: ControlType.Number, title: "Exponent", min: 0, max: 10 },
  };

  constructor (props) {
    super(props);
    this.container = null;
    this.canvas = null;
  }

  componentDidMount () {
    this.canvas = document.createElement('canvas');
    this.container.appendChild(this.canvas);
    this.manager = null;

    canvasSketch(sketch, {
      ...settings,
      data: this.props,
      scaleToFit: false,
      dimensions: [ this.props.width, this.props.height ],
      canvas: this.canvas
    }).then(manager => {
      this.manager = manager;
      this.sendProps(this.props);
    });
  }

  componentWillReceiveProps (props) {
    this.sendProps(props);
  }

  sendProps (props) {
    if (this.manager) {
      if (props.playing) this.manager.play();
      else this.manager.pause();
      this.manager.update({
        data: props,
        dimensions: [ props.width, props.height ]
      });
    }
  }

  componentWillUnmount () {
    if (this.manager) {
      this.manager.unload();
    }
    if (this.canvas) {
      this.canvas.parentElement.removeChild(this.canvas);
    }
  }

  render() {
    return <div ref={c => {this.container = c; }} />;
  }
}
