// @framer/mattdesl.motiongrid/code/sketch.js
const lerp = require('lerp');

export const settings = {
  animate: true,
  duration: 3,
  dimensions: [ 640, 640 ],
  scaleToView: true,
  playbackRate: 'throttle',
  fps: 24
};

// Start the sketch
export default () => {
  return ({ context, frame, width, height, playhead, settings }) => {
    const {
      background,
      foreground,
      gridSize,
      padding,
      thickness,
      length,
      rotation,
      stroke,
      uMod,
      vMod,
      exponential
    } = settings.data;

    context.clearRect(0, 0, width, height);
    context.fillStyle = background;
    context.fillRect(0, 0, width, height);

    const paddingAmount = width * (padding / 100);
    const tileSize = (width - paddingAmount * 2) / gridSize;

    for (let x = 0; x < gridSize; x++) {
      for (let y = 0; y < gridSize; y++) {
        // get a 0..1 UV coordinate
        const u = gridSize <= 1 ? 0.5 : x / (gridSize - 1);
        const v = gridSize <= 1 ? 0.5 : y / (gridSize - 1);

        // scale to dimensions with a border padding
        const tx = lerp(paddingAmount, width - paddingAmount, u);
        const ty = lerp(paddingAmount, height - paddingAmount, v);

        // here we get a 't' value between 0..1 that
        // shifts subtly across the UV coordinates
        const offset = u * (uMod / 100) + v * (vMod / 100);
        const t = (playhead + offset) % 1;

        // now we get a value that varies from 0..1 and back
        let mod = Math.sin(t * Math.PI);

        // we make it 'ease' a bit more dramatically with exponential
        if (exponential !== 0) mod = Math.pow(mod, exponential);

        // now choose a length, thickness and initial rotation
        const lengthAmount = tileSize * (length / 100);
        const thicknessAmount = tileSize * (thickness / 100);
        const initialRotation = rotation * Math.PI / 180;

        // And rotate each line a bit by our modifier
        const rotationRad = initialRotation + mod * Math.PI;

        // Now render...
        draw(context, stroke, foreground, tx, ty, lengthAmount, thicknessAmount, rotationRad);
      }
    }
  };

  function draw (context, stroke, foreground, x, y, length, thickness, rotation) {
    context.save();
    context.fillStyle = context.strokeStyle = foreground;

    // Rotate in place
    context.translate(x, y);
    context.rotate(rotation);
    context.translate(-x, -y);

    // Draw the line
    context.beginPath();
    context.rect(x - length / 2, y - thickness / 2, length, thickness);
    if (stroke) context.stroke();
    else context.fill();
    context.restore();
  }
};
