// @framer/maxsteitle.resizable-window/code/headerStyles.tsx
import * as React from "react"
import { Frame } from "framer"

export function safariHeader(parent, appearance, title, style) {
    return (
        <Frame
            style={{
                height: 38,
                background: appearance
                    ? "linear-gradient(rgba(33,33,33,.6), rgba(10,10,10,.6)), linear-gradient(rgba(255,255,255,.5), rgba(255,255,255,.5))"
                    : "linear-gradient(rgba(230,230,230,.75), rgba(195,195,195,.75))",
                color: appearance
                    ? "rgba(255,255,255,.85)"
                    : "rgba(0,0,0,0.73)",
                boxShadow: appearance
                    ? " inset 0 -1px rgba(0,0,0,0.2)"
                    : "inset 0 -1px 1px rgba(0,0,0,0.15), inset 0 1px rgba(255,255,255,0.4)",
                borderRadius: "6px 6px 0 0",
                width: "100%",
                WebkitBackdropFilter: "blur(24px)",
                overflow: "hidden",
            }}
            onPan={parent.panHeaderHandler}
            onPanStart={parent.panHeaderStartHandler}
            top={0}
            left={0}
        >
            <div
                style={{
                    width: "100%",
                    height: 38,
                    display: "flex",
                    alignItems: "center",
                    justifyContent: "space-between",
                    whiteSpace: "nowrap",
                    textAlign: "left",
                }}
            >
                <div style={{ display: "flex", alignItems: "center" }}>
                    <div style={{ ...trafficLightsWrapper, marginLeft: 14 }}>
                        <div
                            style={trafficLight(
                                "#FF6157",
                                "#E24640",
                                appearance,
                                style
                            )}
                        />
                        <div
                            style={trafficLight(
                                "#FFC12F",
                                "#DFA023",
                                appearance,
                                style
                            )}
                        />
                        <div
                            style={trafficLight(
                                "#2ACB42",
                                "#1BAC2C",
                                appearance,
                                style
                            )}
                        />
                    </div>
                    <div
                        style={{
                            flexShrink: 0,
                            width: 66,
                            display: "flex",
                            paddingLeft: 13,
                        }}
                    >
                        <div
                            style={{
                                width: 26,
                                height: 22,
                                borderRadius: 4,
                                background: appearance
                                    ? "linear-gradient(rgba(255,255,255,.25), rgba(255,255,255,.25))"
                                    : "linear-gradient(#FEFEFE, #F1F1F1)",
                                boxShadow:
                                    "0 1px 1px rgba(0,0,0,0.08), 0 0 0 1px rgba(0,0,0,0.08)",
                                display: "flex",
                                justifyContent: "center",
                                alignItems: "center",
                            }}
                        >
                            <svg
                                width="14"
                                height="14"
                                viewBox="0 0 14 14"
                                fill="none"
                                xmlns="http://www.w3.org/2000/svg"
                            >
                                <path
                                    d="M8.75137 12.9187C8.46959 13.2005 8.01274 13.2005 7.73096 12.9187L2.25156 7.43928C2.04481 7.23253 2.04481 6.89733 2.25156 6.69058L7.86084 1.0813C8.14262 0.79952 8.59947 0.79952 8.88125 1.0813C9.16303 1.36308 9.16303 1.81993 8.88125 2.10171L3.91802 7.06493L8.75137 11.8983C9.03315 12.1801 9.03315 12.6369 8.75137 12.9187Z"
                                    fill={appearance ? "#FFFFFF" : "#808080"}
                                />
                            </svg>
                        </div>
                        <div
                            style={{
                                width: 26,
                                height: 22,
                                background: appearance
                                    ? "linear-gradient(rgba(255,255,255,.25), rgba(255,255,255,.25))"
                                    : "linear-gradient(#FEFEFE, #F1F1F1)",
                                marginLeft: 1,
                                borderRadius: 4,
                                boxShadow:
                                    "0 1px 1px rgba(0,0,0,0.08), 0 0 0 1px rgba(0,0,0,0.08)",
                                display: "flex",
                                justifyContent: "center",
                                alignItems: "center",
                            }}
                        >
                            <svg
                                width="14"
                                height="14"
                                viewBox="0 0 14 14"
                                fill="none"
                                xmlns="http://www.w3.org/2000/svg"
                            >
                                <path
                                    d="M4.38748 12.9187C4.66925 13.2005 5.12611 13.2005 5.40788 12.9187L10.5545 7.77204C10.9451 7.38151 10.9451 6.74835 10.5545 6.35783L5.27801 1.0813C4.99623 0.79952 4.53938 0.79952 4.2576 1.0813C3.97582 1.36308 3.97582 1.81993 4.2576 2.10171L9.22082 7.06493L4.38747 11.8983C4.1057 12.1801 4.1057 12.6369 4.38748 12.9187Z"
                                    fill={appearance ? "#FFFFFF" : "#808080"}
                                />
                            </svg>
                        </div>
                    </div>
                </div>

                <div
                    style={{
                        height: 22,
                        width: "40%",
                        borderRadius: 4,
                        background: appearance
                            ? "linear-gradient(rgba(255,255,255,.2), rgba(255,255,255,.2))"
                            : "linear-gradient(#FEFEFE, #F1F1F1)",
                        boxShadow:
                            "0 1px 1px rgba(0,0,0,0.08), 0 0 0 1px rgba(0,0,0,0.08)",
                        display: "flex",
                        flexDirection: "row",
                        fontSize: 13,
                        justifyContent: "space-between",
                        alignItems: "center",
                        color: appearance ? "white" : "black",
                    }}
                >
                    <div style={spacer(15)} />
                    <div style={{ display: "flex" }}>
                        <svg
                            width="15"
                            height="15"
                            viewBox="1 -1 15 15"
                            fill="none"
                            xmlns="http://www.w3.org/2000/svg"
                        >
                            <path
                                fillRule="evenodd"
                                clipRule="evenodd"
                                d="M4.50013 4L4.50004 5.5H3V12.5H12V5.5H10.5V4C10.5 2.65431 9.20104 1.5 7.49987 1.5C5.79861 1.5 4.50016 2.65442 4.50013 4ZM7.49987 2.5C6.20089 2.5 5.50013 3.3456 5.50013 4L5.50004 5.5H9.5V4C9.5 3.34569 8.79896 2.5 7.49987 2.5Z"
                                fill={appearance ? "white" : "black"}
                            />
                        </svg>

                        {title}
                    </div>

                    <div style={{ paddingRight: "3px" }}>
                        <svg
                            width="15"
                            height="15"
                            viewBox="0 -2 15 15"
                            fill="none"
                            xmlns="http://www.w3.org/2000/svg"
                        >
                            <path
                                fillRule="evenodd"
                                clipRule="evenodd"
                                d="M7.5 5.5L12 2.75L7.5 0V2C4.46243 2 2 4.46243 2 7.5C2 10.5376 4.46243 13 7.5 13C10.5376 13 13 10.5376 13 7.5H12C12 9.98528 9.98528 12 7.5 12C5.01472 12 3 9.98528 3 7.5C3 5.01472 5.01472 3 7.5 3V5.5Z"
                                fill={appearance ? "#FFFFFF" : "#7E7E7E"}
                            />
                        </svg>
                    </div>
                </div>
                <div
                    style={{
                        flexShrink: 0,
                        display: "flex",
                        alignItems: "flex-end",
                        height: 38,
                    }}
                >
                    <div style={spacer(106)} />
                    <div
                        style={{
                            width: 26,
                            height: 26,
                            background:
                                "linear-gradient(rgba(50,50,50,.2), rgba(30,30,30,.2))",
                            boxShadow: appearance
                                ? "inset 1px 1px 0 rgba(255,255,255,.22)"
                                : "inset 1px 1px 0 rgba(0,0,0,0.08)",
                            flexShrink: 0,
                            alignSelf: "flex-end",
                        }}
                    >
                        <div style={{ left: 7, top: 8, position: "relative" }}>
                            <svg
                                width="11"
                                height="11"
                                viewBox="0 0 11 11"
                                fill="none"
                                xmlns="http://www.w3.org/2000/svg"
                            >
                                <path
                                    fillRule="evenodd"
                                    clipRule="evenodd"
                                    d="M6 0H5V5H0V6H5V11H6V6H11V5H6V0Z"
                                    fill={
                                        appearance
                                            ? "rgba(255,255,255,0.6)"
                                            : "rgba(0,0,0,0.5)"
                                    }
                                />
                            </svg>
                        </div>
                    </div>
                </div>
            </div>
        </Frame>
    )
}

export function macOSHeader(
    parent: any,
    appearance: boolean,
    title: string,
    style
) {
    return (
        <Frame
            style={{
                height: 22,
                background: appearance
                    ? "linear-gradient(#4B4B4B, #3D3D3D)"
                    : "linear-gradient(#EDECED, #D2D1D2)",
                boxShadow: appearance
                    ? "inset 0 -1px rgba(0,0,0,1), inset 0 -1px rgba(43,43,43,1), inset 0 -1px rgba(255,255,255,0.01)"
                    : "inset 0 1px 1px rgba(255,255,255,0.5), inset 0 -1px rgba(183,180,183,1), inset 0 -1px rgba(183,180,183,0.3)",
                borderRadius: "4px 4px 0 0",
                color: appearance
                    ? "rgba(255,255,255,.85)"
                    : "rgba(0,0,0,0.73)",
                width: "100%",
            }}
            onPan={parent.panHeaderHandler}
            onPanStart={parent.panHeaderStartHandler}
            top={0}
            left={0}
        >
            <div
                style={{
                    width: "100%",
                    height: 22,
                    display: "flex",
                    alignItems: "center",
                    justifyContent: "space-between",
                    whiteSpace: "nowrap",
                    textAlign: "left",
                    padding: 8,
                    overflow: "hidden",
                }}
            >
                <div style={trafficLightsWrapper}>
                    <div
                        style={trafficLight(
                            "#FF6157",
                            "#E24640",
                            appearance,
                            style
                        )}
                    />
                    <div
                        style={trafficLight(
                            "#FFC12F",
                            "#DFA023",
                            appearance,
                            style
                        )}
                    />
                    <div
                        style={trafficLight(
                            "#2ACB42",
                            "#1BAC2C",
                            appearance,
                            style
                        )}
                    />
                </div>
                <div style={{ fontSize: 13 }}>{title}</div>
                <div style={spacer(52)} />
            </div>
        </Frame>
    )
}

export function noHeader(parent: any, height: number) {
    return (
        <Frame
            style={{
                height: height,
                background: null,
                borderRadius: "4px 4px 0 0",
                width: "100%",
                paddingTop: 0,
            }}
            onPan={parent.panHeaderHandler}
            onPanStart={parent.panHeaderStartHandler}
            top={0}
            left={0}
        />
    )
}

const trafficLightsWrapper: React.CSSProperties = {
    width: 52,
    flexShrink: 0,
    height: "100%",
    display: "flex",
    alignItems: "center",
    justifyContent: "space-between",
}

function trafficLight(
    background: string,
    borderColor: string,
    appearance: boolean,
    style: string
): React.CSSProperties {
    return {
        width: 12,
        height: 12,
        background,
        borderRadius: "50%",
        borderColor: borderColor,
        borderStyle: "solid",
        borderWidth: appearance
            ? style === "macOS" || style === "Safari" || style === "Chrome"
                ? 0
                : 0.5
            : 0.5,
    }
}

function spacer(width: number): React.CSSProperties {
    return {
        width: width,
    }
}

export function firefoxHeader(
    parent: any,
    appearance: boolean,
    title: string,
    style: string
) {
    return (
        <Frame
            top={0}
            left={0}
            onPan={parent.panHeaderHandler}
            onPanStart={parent.panHeaderStartHandler}
            style={{
                width: "100%",
                height: 74,
                borderRadius: "6px 6px 0px 0px",
            }}
            overflow={"hidden"}
            background={null}
        >
            <div
                style={{
                    height: 33,
                    width: "100%",
                    background: "rgba(0,0,0,0.6)",
                    WebkitBackdropFilter: "blur(24px)",
                    display: "flex",
                    paddingLeft: 13,
                    borderRadius: "6px 6px 0px 0px",
                    boxShadow: "inset 0 -1px 0 rgba(0,0,0,0.2)",
                }}
            >
                <div style={trafficLightsWrapper}>
                    <div
                        style={trafficLight(
                            "#FF6157",
                            "#E24640",
                            appearance,
                            style
                        )}
                    />
                    <div
                        style={trafficLight(
                            "#FFC12F",
                            "#DFA023",
                            appearance,
                            style
                        )}
                    />
                    <div
                        style={trafficLight(
                            "#2ACB42",
                            "#1BAC2C",
                            appearance,
                            style
                        )}
                    />
                </div>
                <div
                    style={{
                        height: "100%",
                        width: 224,
                        // flexShrink: 0,
                        background: "#F9F9FA",
                        marginLeft: 41,
                        boxShadow:
                            "inset 0 2px 0 #0082FF, 1px 0 0 rgba(0,0,0,0.1), -1px 0 0 rgba(0,0,0,0.1)",
                    }}
                >
                    <div
                        style={{
                            height: 24,
                            margin: 6,
                            padding: 4,
                            display: "flex",
                        }}
                    >
                        <div
                            style={{
                                height: 16,
                                width: 16,
                                marginTop: -1,
                                flexShrink: 0,
                                marginLeft: -1,
                            }}
                        >
                            <img
                                style={{ height: 16, width: 16 }}
                                src={`https://api.faviconkit.com/${title}/32`}
                            />
                        </div>
                        <div
                            style={{
                                flexGrow: 1,
                                height: 16,
                                fontSize: 11,
                                color: "#17191A",
                                lineHeight: "15px",
                                paddingLeft: 6,
                                overflow: "hidden",
                            }}
                        >
                            {title
                                .substring(0, title.lastIndexOf("."))
                                .replace(/^\w/, c => c.toUpperCase())}
                        </div>
                        <div
                            style={{
                                flexShrink: 0,
                                width: 14,
                                height: 16,
                                marginTop: 1,
                                paddingLeft: 1,
                            }}
                        >
                            <svg
                                width="12"
                                height="12"
                                viewBox="0 0 12 12"
                                fill="none"
                                xmlns="http://www.w3.org/2000/svg"
                            >
                                <path
                                    fillRule="evenodd"
                                    clipRule="evenodd"
                                    d="M0.969365 2.03027C0.77474 1.84196 0.696687 1.56311 0.765203 1.30089C0.833719 1.03867 1.03814 0.833882 1.29991 0.765246C1.56167 0.696609 1.84003 0.774801 2.028 0.969769L5.99078 4.93952L9.95356 0.969769C10.1415 0.774801 10.4199 0.696609 10.6817 0.765246C10.9434 0.833882 11.1478 1.03867 11.2164 1.30089C11.2849 1.56311 11.2068 1.84196 11.0122 2.03027L7.04942 6.00002L11.0122 9.96977C11.2959 10.2641 11.2919 10.7318 11.0031 11.0211C10.7143 11.3104 10.2473 11.3145 9.95356 11.0303L5.99078 7.06052L2.028 11.0303C1.73424 11.3145 1.26728 11.3104 0.978486 11.0211C0.689694 10.7318 0.685635 10.2641 0.969365 9.96977L4.93215 6.00002L0.969365 2.03027Z"
                                    fill="#5A5B5C"
                                />
                            </svg>
                        </div>
                    </div>
                </div>
                <div
                    style={{
                        height: 18,
                        width: 14,
                        flexShrink: 0,
                        // background: 'red',
                        marginTop: 9,
                        marginLeft: 10,
                    }}
                >
                    <svg
                        width="14"
                        height="14"
                        viewBox="0 0 14 14"
                        fill="none"
                        xmlns="http://www.w3.org/2000/svg"
                    >
                        <path
                            d="M8 6H13C13.5523 6 14 6.44772 14 7C14 7.55228 13.5523 8 13 8H8V13C8 13.5523 7.55228 14 7 14C6.44772 14 6 13.5523 6 13V8H1C0.447715 8 0 7.55228 0 7C0 6.44772 0.447715 6 1 6H6V1C6 0.447715 6.44772 0 7 0C7.55228 0 8 0.447715 8 1V6Z"
                            fill="#BCBDC1"
                        />
                    </svg>
                </div>
                <div style={spacer(12)} />
            </div>
            <div
                style={{
                    height: 41,
                    width: "100%",
                    background: "#F9F9FA",
                    display: "flex",
                    justifyContent: "space-between",
                    boxShadow: "inset 0 -1px 0 #E1E1E2",
                }}
            >
                <div
                    style={{
                        height: 34,
                        width: 92,
                        marginLeft: 3,
                        paddingTop: 3,
                        flexShrink: 0,
                        display: "flex",
                    }}
                >
                    <div
                        style={{
                            height: 34,
                            width: 34,
                            borderRadius: "50%",
                            border: "1px solid #B1B1B2",
                            background: "#FEFEFE",
                            flexShrink: 0,
                        }}
                    >
                        <div style={{ marginLeft: 8, marginTop: 9 }}>
                            <svg
                                width="16"
                                height="14"
                                viewBox="0 0 16 14"
                                fill="none"
                                xmlns="http://www.w3.org/2000/svg"
                            >
                                <path
                                    fillRule="evenodd"
                                    clipRule="evenodd"
                                    d="M7.70679 0.293048C8.09717 0.683897 8.09717 1.31746 7.70679 1.70831L3.41379 6.00515H14.9998C15.5521 6.00515 15.9998 6.45327 15.9998 7.00604C15.9998 7.55882 15.5521 8.00694 14.9998 8.00694H3.41379L7.70679 12.3038C8.08576 12.6965 8.08034 13.3208 7.6946 13.7068C7.30887 14.0929 6.68517 14.0984 6.29279 13.719L0.292786 7.71368C-0.0975955 7.32283 -0.0975955 6.68926 0.292786 6.29841L6.29279 0.293048C6.68329 -0.0976827 7.31629 -0.0976827 7.70679 0.293048Z"
                                    fill="#5C5D5E"
                                />
                            </svg>
                        </div>
                    </div>
                    <div
                        style={{
                            width: 34,
                            height: 34,
                            flexShrink: 0,
                        }}
                    >
                        <div style={{ marginLeft: 8, marginTop: 10 }}>
                            <svg
                                width="16"
                                height="14"
                                viewBox="0 0 16 14"
                                fill="none"
                                xmlns="http://www.w3.org/2000/svg"
                            >
                                <path
                                    fillRule="evenodd"
                                    clipRule="evenodd"
                                    d="M8.293 0.293048C8.6835 -0.0976827 9.3165 -0.0976827 9.707 0.293048L15.707 6.29841C16.0974 6.68926 16.0974 7.32283 15.707 7.71368L9.707 13.719C9.31462 14.0984 8.69092 14.0929 8.30518 13.7068C7.91945 13.3208 7.91403 12.6965 8.293 12.3038L12.586 8.00694H1C0.447715 8.00694 0 7.55882 0 7.00604C0 6.45327 0.447715 6.00515 1 6.00515H12.586L8.293 1.70831C7.90262 1.31746 7.90262 0.683897 8.293 0.293048Z"
                                    fill="#B6B6B8"
                                />
                            </svg>
                        </div>
                    </div>
                    <div
                        style={{
                            height: 34,
                            width: 28,
                            justifyContent: "center",
                            flexShrink: 0,
                        }}
                    >
                        <div style={{ marginLeft: 6, marginTop: 9 }}>
                            <svg
                                width="16"
                                height="15"
                                viewBox="0 0 16 15"
                                fill="none"
                                xmlns="http://www.w3.org/2000/svg"
                            >
                                <path
                                    d="M12.5831 6.00002C11.5829 3.70023 8.99374 2.46775 6.53817 3.2185C5.71057 3.47152 4.99635 3.91979 4.4306 4.49735C4.42174 4.50728 4.41265 4.51708 4.40332 4.52673C4.27776 4.65676 4.16065 4.79184 4.05199 4.93129C3.61711 5.49177 3.30423 6.14063 3.13741 6.83261C2.97189 7.5249 2.9561 8.24509 3.08912 8.94192C3.12264 9.1155 3.16565 9.28902 3.2185 9.46188C3.22243 9.47472 3.22608 9.48757 3.22946 9.50044C3.47135 10.2719 3.90421 10.9956 4.52674 11.5967C6.51314 13.515 9.67847 13.4597 11.5967 11.4733C11.9804 11.076 12.6134 11.065 13.0107 11.4486C13.408 11.8323 13.4191 12.4654 13.0354 12.8626C10.3499 15.6436 5.91838 15.7209 3.13742 13.0354C2.05824 11.9932 1.3862 10.6882 1.12476 9.31867C0.931198 8.31609 0.964195 7.31206 1.19236 6.36565C1.41873 5.4188 1.84515 4.50922 2.47277 3.70379C3.32746 2.60224 4.51874 1.74452 5.95343 1.30589C9.0909 0.34667 12.3841 1.70003 14 4.39127V2.00002C14 1.44774 14.4477 1.00002 15 1.00002C15.5523 1.00002 16 1.44774 16 2.00002V7.00002C16 7.55231 15.5523 8.00002 15 8.00002H10C9.44774 8.00002 9.00003 7.55231 9.00003 7.00002C9.00003 6.44774 9.44774 6.00002 10 6.00002H12.5831Z"
                                    fill="#595A5C"
                                />
                            </svg>
                        </div>
                    </div>
                </div>
                <div
                    style={{
                        height: 30,
                        width: "66.3%",
                        background: "white",
                        marginTop: 5,
                        boxShadow: "inset 0 0 0 1px #BFC2C3",
                        borderRadius: 4,
                        display: "flex",
                        paddingLeft: 8,
                        paddingRight: 8,
                        alignItems: "center",
                        fontSize: 14,
                        fontWeight: 400,
                        overflow: "hidden",
                    }}
                >
                    <div
                        style={{ paddingRight: 8, marginTop: 2, flexShrink: 0 }}
                    >
                        <svg
                            width="12"
                            height="14"
                            viewBox="0 0 12 14"
                            fill="none"
                            xmlns="http://www.w3.org/2000/svg"
                        >
                            <path
                                d="M2 6V4C2 1.79086 3.79086 0 6 0C8.20914 0 10 1.79086 10 4V6H10.5C11.3284 6 12 6.67157 12 7.5V12.5C12 13.3284 11.3284 14 10.5 14H1.5C0.671573 14 0 13.3284 0 12.5V7.5C0 6.67157 0.671573 6 1.5 6H2ZM4 6H8V4C8 2.89543 7.10457 2 6 2C4.89543 2 4 2.89543 4 4V6Z"
                                fill="#00C000"
                            />
                        </svg>
                    </div>
                    <div style={{ flexGrow: 1 }}>{title}</div>

                    <div
                        style={{
                            width: 16,
                            height: 16,
                            marginTop: -7,
                            flexShrink: 0,
                        }}
                    >
                        <svg
                            width="16"
                            height="4"
                            viewBox="0 0 16 4"
                            fill="none"
                            xmlns="http://www.w3.org/2000/svg"
                        >
                            <path
                                fillRule="evenodd"
                                clipRule="evenodd"
                                d="M4 2C4 3.10457 3.10457 4 2 4C0.89543 4 0 3.10457 0 2C0 0.89543 0.89543 0 2 0C3.10457 0 4 0.89543 4 2ZM10 2C10 3.10457 9.10457 4 8 4C6.89543 4 6 3.10457 6 2C6 0.89543 6.89543 0 8 0C9.10457 0 10 0.89543 10 2ZM14 4C15.1046 4 16 3.10457 16 2C16 0.89543 15.1046 0 14 0C12.8954 0 12 0.89543 12 2C12 3.10457 12.8954 4 14 4Z"
                                fill="#666666"
                            />
                        </svg>
                    </div>
                </div>
                <div
                    style={{
                        height: 34,
                        width: 39,
                        alignSelf: "center",
                        flexShrink: 0,
                        display: "flex",
                        justifyContent: "center",
                        borderLeft: "1px solid #DFDFE0",
                    }}
                >
                    <div style={{ marginTop: 10.5 }}>
                        <svg
                            width="12"
                            height="12"
                            viewBox="0 0 12 12"
                            fill="none"
                            xmlns="http://www.w3.org/2000/svg"
                        >
                            <path
                                fillRule="evenodd"
                                clipRule="evenodd"
                                d="M0 1C0 0.447715 0.447715 0 1 0H11C11.5523 0 12 0.447715 12 1C12 1.55228 11.5523 2 11 2H1C0.447715 2 0 1.55228 0 1ZM0 6C0 5.44772 0.447715 5 1 5H11C11.5523 5 12 5.44772 12 6C12 6.55228 11.5523 7 11 7H1C0.447715 7 0 6.55228 0 6ZM1 10C0.447715 10 0 10.4477 0 11C0 11.5523 0.447715 12 1 12H11C11.5523 12 12 11.5523 12 11C12 10.4477 11.5523 10 11 10H1Z"
                                fill="#4A4A4B"
                            />
                        </svg>
                    </div>
                </div>
            </div>
        </Frame>
    )
}

export function chromeHeader(
    parent: any,
    appearance: boolean,
    title: string,
    style: string
) {
    return (
        <Frame
            top={0}
            left={0}
            onPan={parent.panHeaderHandler}
            onPanStart={parent.panHeaderStartHandler}
            style={{
                width: "100%",
                height: 79,
                borderRadius: "6px 6px 0px 0px",
            }}
            overflow={"hidden"}
            background={null}
        >
            <div
                style={{
                    height: 42,
                    width: "100%",
                    background: appearance ? "#202124" : "#DDE1E5",
                    display: "flex",
                    paddingLeft: 13,
                    borderRadius: "6px 6px 0px 0px",
                    boxShadow: appearance
                        ? null
                        : "inset 0 1px 1px rgba(255,255,255,0.35)",
                }}
            >
                <div style={trafficLightsWrapper}>
                    <div
                        style={trafficLight(
                            "#FF6157",
                            "#E24640",
                            appearance,
                            style
                        )}
                    />
                    <div
                        style={trafficLight(
                            "#FFC12F",
                            "#DFA023",
                            appearance,
                            style
                        )}
                    />
                    <div
                        style={trafficLight(
                            "#2ACB42",
                            "#1BAC2C",
                            appearance,
                            style
                        )}
                    />
                </div>
                <div
                    style={{
                        height: 34,
                        alignSelf: "flex-end",
                        width: 240,
                        background: appearance ? "#323639" : "white",
                        marginLeft: 13,
                        borderRadius: "8px 8px 0 0",
                    }}
                >
                    <div
                        style={{
                            height: 24,
                            margin: 5,
                            padding: 4,
                            display: "flex",
                        }}
                    >
                        <div
                            style={{
                                marginLeft: 3,
                                height: 16,
                                width: 16,
                                flexShrink: 0,
                            }}
                        >
                            <img
                                style={{ height: 16, width: 16 }}
                                src={`https://api.faviconkit.com/${title}/32`}
                            />
                        </div>
                        <div
                            style={{
                                flexGrow: 1,
                                height: 16,
                                fontSize: 13,
                                marginLeft: 3,
                                color: appearance ? "#F0F2F3" : "#3C4043",
                                lineHeight: "15px",
                                paddingLeft: 5,
                                overflow: "hidden",
                            }}
                        >
                            {title
                                .substring(0, title.lastIndexOf("."))
                                .replace(/^\w/, c => c.toUpperCase())}
                        </div>
                        <div
                            style={{
                                flexShrink: 0,
                                width: 14,
                                height: 16,
                                marginTop: 1,
                                paddingLeft: 3,
                            }}
                        >
                            <svg
                                width="8"
                                height="8"
                                viewBox="0 0 8 8"
                                fill="none"
                                xmlns="http://www.w3.org/2000/svg"
                            >
                                <path
                                    d="M7.66101 1.26501L4.82601 4.10001L7.66101 6.93501C7.8891 7.16311 7.8891 7.53292 7.66101 7.76101C7.43291 7.98911 7.0631 7.98911 6.83501 7.76101L4.00001 4.92601L1.16501 7.76101C0.936915 7.98911 0.567102 7.98911 0.339008 7.76101C0.110915 7.53292 0.110915 7.16311 0.339008 6.93501L3.17401 4.10001L0.339008 1.26501C0.110915 1.03692 0.110915 0.667108 0.339008 0.439015C0.567102 0.210921 0.936915 0.210921 1.16501 0.439015L4.00001 3.27401L6.83501 0.439015C7.0631 0.210921 7.43291 0.210921 7.66101 0.439015C7.8891 0.667108 7.8891 1.03692 7.66101 1.26501Z"
                                    fill={appearance ? "#B6B8BA" : "#5F6367"}
                                />
                            </svg>
                        </div>
                    </div>
                </div>
                <div style={{ position: "absolute", left: 70, top: 34 }}>
                    <div
                        style={{
                            height: 8,
                            width: 8,
                            background: appearance ? "#323639" : "white",
                            position: "absolute",
                        }}
                    />
                    <div
                        style={{
                            height: 8,
                            width: 8,
                            background: appearance ? "#202124" : "#DDE1E5",
                            borderRadius: "0 0 8px 0",
                            position: "absolute",
                        }}
                    />
                </div>
                <div style={{ marginTop: 34 }}>
                    <div
                        style={{
                            height: 8,
                            width: 8,
                            background: appearance ? "#323639" : "white",
                            position: "absolute",
                        }}
                    />
                    <div
                        style={{
                            height: 8,
                            width: 8,
                            background: appearance ? "#202124" : "#DDE1E5",
                            borderRadius: "0 0 0 8px",
                            position: "absolute",
                        }}
                    />
                </div>
                <div
                    style={{
                        height: 16,
                        width: 16,
                        flexShrink: 0,
                        alignSelf: "flex-end",
                        marginLeft: 14,
                        marginBottom: 9,
                    }}
                >
                    <div style={{ marginTop: 2, marginLeft: 2 }}>
                        <svg
                            width="12"
                            height="12"
                            viewBox="0 0 12 12"
                            fill="none"
                            xmlns="http://www.w3.org/2000/svg"
                        >
                            <path
                                d="M11 6.82005H6.81999V11.0001C6.81999 11.4529 6.45287 11.8201 5.99999 11.8201C5.54712 11.8201 5.17999 11.4529 5.17999 11.0001V6.82005H0.999993C0.547119 6.82005 0.179993 6.45293 0.179993 6.00005C0.179993 5.54718 0.547119 5.18005 0.999993 5.18005H5.17999V1.00005C5.17999 0.54718 5.54712 0.180054 5.99999 0.180054C6.45287 0.180054 6.81999 0.54718 6.81999 1.00005V5.18005H11C11.4529 5.18005 11.82 5.54718 11.82 6.00005C11.82 6.45293 11.4529 6.82005 11 6.82005Z"
                                fill={appearance ? "#BDC1C5" : "#3C4043"}
                            />
                        </svg>
                    </div>
                </div>
                <div style={spacer(52)} />
            </div>
            <div
                style={{
                    height: 37,
                    width: "100%",
                    background: appearance ? "#323639" : "white",
                    display: "flex",
                    justifyContent: "space-between",
                    borderBottom: "0.5px solid #B6B4B6",
                }}
            >
                <div
                    style={{
                        height: 34,
                        width: 92,
                        marginLeft: 3,
                        paddingTop: 3,
                        flexShrink: 0,
                        display: "flex",
                    }}
                >
                    <div
                        style={{
                            height: 34,
                            width: 34,
                            flexShrink: 0,
                        }}
                    >
                        <div style={{ marginLeft: 11, marginTop: 7 }}>
                            <svg
                                width="16"
                                height="16"
                                viewBox="0 0 16 16"
                                fill="none"
                                xmlns="http://www.w3.org/2000/svg"
                            >
                                <path
                                    d="M14.56 7.98441C14.56 8.43728 14.1929 8.80441 13.74 8.80441H4.40001L8.47128 12.8931C8.79663 13.2199 8.79606 13.7484 8.47001 14.0744C8.14416 14.4003 7.61586 14.4003 7.29001 14.0744L1.20001 7.98441L7.29001 1.89441C7.61586 1.56856 8.14416 1.56856 8.47001 1.89441C8.79606 2.22046 8.79663 2.74892 8.47128 3.07567L4.40001 7.16441H13.74C14.1929 7.16441 14.56 7.53154 14.56 7.98441Z"
                                    fill={appearance ? "white" : "#5F6367"}
                                />
                            </svg>
                        </div>
                    </div>
                    <div
                        style={{
                            width: 34,
                            height: 34,
                            flexShrink: 0,
                        }}
                    >
                        <div style={{ marginLeft: 9, marginTop: 7 }}>
                            <svg
                                width="16"
                                height="16"
                                viewBox="0 0 16 16"
                                fill="none"
                                xmlns="http://www.w3.org/2000/svg"
                            >
                                <path
                                    d="M1.43999 7.98441C1.43999 8.43728 1.80711 8.80441 2.25999 8.80441H11.6L7.52872 12.8931C7.20337 13.2199 7.20394 13.7484 7.52999 14.0744C7.85584 14.4003 8.38414 14.4003 8.70999 14.0744L14.8 7.98441L8.70999 1.89441C8.38414 1.56856 7.85584 1.56856 7.52999 1.89441C7.20394 2.22046 7.20337 2.74892 7.52872 3.07567L11.6 7.16441H2.25999C1.80711 7.16441 1.43999 7.53154 1.43999 7.98441Z"
                                    fill={appearance ? "#898D8E" : "#BABCBE"}
                                />
                            </svg>
                        </div>
                    </div>
                    <div
                        style={{
                            height: 34,
                            width: 28,
                            flexShrink: 0,
                        }}
                    >
                        <div style={{ marginLeft: 7, marginTop: 6.5 }}>
                            <svg
                                width="16"
                                height="16"
                                viewBox="0 0 16 16"
                                fill="none"
                                xmlns="http://www.w3.org/2000/svg"
                            >
                                <path
                                    d="M12.7 3.27995L14.66 1.31995V7.17995H8.8L11.5 4.47995C11.0333 4.01328 10.4967 3.64995 9.89 3.38995C9.28333 3.12994 8.64667 2.99995 7.98 2.99995C7.07332 2.99995 6.23667 3.22328 5.47 3.66995C4.70333 4.11662 4.09667 4.72328 3.65 5.48995C3.20333 6.25662 2.98 7.09327 2.98 7.99995C2.98 8.90662 3.20333 9.74328 3.65 10.5099C4.09667 11.2766 4.70333 11.8833 5.47 12.3299C6.23667 12.7766 7.07332 12.9999 7.98 12.9999C9.06 12.9999 10.0333 12.6933 10.9 12.0799C11.7667 11.4666 12.3667 10.6666 12.7 9.67995H14.42C14.18 10.64 13.7467 11.4966 13.12 12.2499C12.4933 13.0033 11.7467 13.5933 10.88 14.0199C9.97333 14.4599 9.00667 14.6799 7.98 14.6799C6.77999 14.6799 5.66 14.3733 4.62 13.7599C3.61999 13.1733 2.82667 12.3733 2.24 11.3599C1.63999 10.3333 1.34 9.21329 1.34 7.99995C1.34 6.78661 1.63999 5.66662 2.24 4.63995C2.82667 3.62661 3.61999 2.82662 4.62 2.23995C5.66 1.62661 6.77999 1.31995 7.98 1.31995C8.88667 1.31995 9.74666 1.49328 10.56 1.83995C11.3733 2.18661 12.0867 2.66661 12.7 3.27995Z"
                                    fill={appearance ? "white" : "#5F6367"}
                                />
                            </svg>
                        </div>
                    </div>
                </div>
                <div
                    style={{
                        height: 28,
                        width: "100%",
                        marginLeft: 13,
                        marginTop: 4,
                        background: appearance ? "#282C2F" : "#EDEFF0",
                        borderRadius: 20,
                        display: "flex",
                    }}
                >
                    <div
                        style={{
                            width: 28,
                            height: 28,
                            paddingLeft: 12,
                            paddingTop: 8,
                            flexShrink: 0,
                        }}
                    >
                        <svg
                            width="12"
                            height="12"
                            viewBox="0 0 12 12"
                            fill="none"
                            xmlns="http://www.w3.org/2000/svg"
                        >
                            <path
                                d="M7.548 3.80799V2.79999C7.548 2.52799 7.478 2.27199 7.338 2.03199C7.198 1.79199 7.00801 1.60199 6.768 1.46199C6.528 1.32199 6.27201 1.25199 6 1.25199C5.728 1.25199 5.47201 1.32199 5.232 1.46199C4.992 1.60199 4.80201 1.79199 4.662 2.03199C4.522 2.27199 4.452 2.52799 4.452 2.79999V3.80799H7.548ZM9 3.80799C9.27201 3.80799 9.508 3.90399 9.708 4.09599C9.90801 4.28799 10.008 4.51999 10.008 4.79199V9.80799C10.008 10.08 9.90801 10.312 9.708 10.504C9.508 10.696 9.27201 10.792 9 10.792H3C2.728 10.792 2.49201 10.696 2.292 10.504C2.092 10.312 1.992 10.08 1.992 9.80799V4.79199C1.992 4.51999 2.092 4.28799 2.292 4.09599C2.49201 3.90399 2.728 3.80799 3 3.80799H3.492V2.79999C3.492 2.34399 3.604 1.92399 3.828 1.53999C4.05201 1.15599 4.356 0.851993 4.74 0.627992C5.12401 0.403991 5.544 0.291992 6 0.291992C6.45601 0.291992 6.876 0.403991 7.26 0.627992C7.64401 0.851993 7.948 1.15599 8.172 1.53999C8.39601 1.92399 8.508 2.34399 8.508 2.79999V3.80799H9Z"
                                fill={appearance ? "#F1F3F4" : "#5F6367"}
                            />
                        </svg>
                    </div>
                    <div
                        style={{
                            fontSize: 14,
                            paddingTop: 4.5,
                            paddingLeft: 6,
                            flexGrow: 1,
                            color: appearance ? "#F1F3F4" : "black",
                        }}
                    >
                        {title}
                    </div>
                    <div
                        style={{
                            height: 28,
                            width: 28,
                            paddingLeft: 2,
                            paddingTop: 6,
                            flexShrink: 0,
                        }}
                    >
                        <svg
                            width="16"
                            height="15"
                            viewBox="0 0 16 15"
                            fill="none"
                            xmlns="http://www.w3.org/2000/svg"
                        >
                            <path
                                d="M7.99999 10.274L10.808 11.948L10.07 8.76204L12.572 6.56604L9.25999 6.29604L7.99999 3.27204L6.73999 6.29604L3.42799 6.56604L5.92999 8.76204L5.19199 11.948L7.99999 10.274ZM15.488 5.63004L11.42 9.17604L12.644 14.45L7.99999 11.642L3.35599 14.45L4.59799 9.17604L0.511993 5.63004L5.89399 5.16204L7.99999 0.212036L10.106 5.16204L15.488 5.63004Z"
                                fill={appearance ? "#BDC1C5" : "#5F6367"}
                            />
                        </svg>
                    </div>
                </div>
                <div
                    style={{
                        height: 34,
                        width: 76,
                        alignSelf: "center",
                        flexShrink: 0,
                        display: "flex",
                    }}
                >
                    <div
                        style={{
                            width: 28,
                            height: 28,
                            paddingLeft: 13,
                            paddingTop: 8,
                        }}
                    >
                        <svg
                            width="18"
                            height="18"
                            viewBox="0 0 18 18"
                            fill="none"
                            xmlns="http://www.w3.org/2000/svg"
                        >
                            <path
                                d="M8.99999 15.0201C10.0133 15.0201 10.9667 14.7667 11.86 14.2601C12.7267 13.7934 13.44 13.1467 14 12.3201C13.9867 11.8134 13.6733 11.3467 13.06 10.9201C12.5267 10.5601 11.8467 10.2667 11.02 10.0401C10.2733 9.84005 9.59666 9.74005 8.98999 9.74005C8.38332 9.74005 7.70666 9.84005 6.95999 10.0401C6.13332 10.2667 5.46 10.5601 4.93999 10.9201C4.32666 11.3467 4.01333 11.8134 3.99999 12.3201C4.56 13.1467 5.27332 13.7934 6.13999 14.2601C7.04666 14.7667 7.99999 15.0201 8.99999 15.0201ZM8.99999 3.18005C8.55999 3.18005 8.14666 3.29339 7.75999 3.52005C7.37332 3.74672 7.06666 4.05339 6.83999 4.44005C6.61333 4.82672 6.49999 5.24005 6.49999 5.68005C6.49999 6.12006 6.61333 6.53339 6.83999 6.92005C7.06666 7.30672 7.37332 7.61339 7.75999 7.84005C8.14666 8.06672 8.55999 8.18005 8.99999 8.18005C9.43999 8.18005 9.85332 8.06672 10.24 7.84005C10.6267 7.61339 10.9333 7.30672 11.16 6.92005C11.3867 6.53339 11.5 6.12006 11.5 5.68005C11.5 5.24005 11.3867 4.82672 11.16 4.44005C10.9333 4.05339 10.6267 3.74672 10.24 3.52005C9.85332 3.29339 9.43999 3.18005 8.99999 3.18005ZM8.99999 0.680054C10.52 0.680054 11.9267 1.06005 13.22 1.82005C14.4733 2.54006 15.46 3.52671 16.18 4.78005C16.94 6.07339 17.32 7.48005 17.32 9.00005C17.32 10.5201 16.94 11.9267 16.18 13.2201C15.46 14.4734 14.4733 15.4601 13.22 16.1801C11.9267 16.9401 10.52 17.3201 8.99999 17.3201C7.47998 17.3201 6.07333 16.9401 4.77999 16.1801C3.52665 15.4467 2.54 14.4534 1.81999 13.2001C1.05999 11.9067 0.679993 10.5067 0.679993 9.00005C0.679993 7.49338 1.05999 6.09339 1.81999 4.80005C2.55333 3.54671 3.54665 2.55339 4.79999 1.82005C6.09333 1.06005 7.49332 0.680054 8.99999 0.680054Z"
                                fill={appearance ? "white" : "#5F6367"}
                            />
                        </svg>
                    </div>
                    <div
                        style={{
                            width: 28,
                            height: 28,
                            paddingLeft: 24,
                            paddingTop: 10,
                        }}
                    >
                        <svg
                            width="4"
                            height="14"
                            viewBox="0 0 4 14"
                            fill="none"
                            xmlns="http://www.w3.org/2000/svg"
                        >
                            <path
                                d="M2.00001 10.3199C2.45334 10.3199 2.84667 10.4866 3.18001 10.8199C3.51334 11.1533 3.68001 11.5466 3.68001 11.9999C3.68001 12.4533 3.51334 12.8466 3.18001 13.1799C2.84667 13.5133 2.45334 13.6799 2.00001 13.6799C1.54667 13.6799 1.15334 13.5133 0.820007 13.1799C0.486672 12.8466 0.320007 12.4533 0.320007 11.9999C0.320007 11.5466 0.486672 11.1533 0.820007 10.8199C1.15334 10.4866 1.54667 10.3199 2.00001 10.3199ZM2.00001 5.31995C2.45334 5.31995 2.84667 5.48661 3.18001 5.81995C3.51334 6.15328 3.68001 6.54661 3.68001 6.99995C3.68001 7.45328 3.51334 7.84661 3.18001 8.17995C2.84667 8.51328 2.45334 8.67995 2.00001 8.67995C1.54667 8.67995 1.15334 8.51328 0.820007 8.17995C0.486672 7.84661 0.320007 7.45328 0.320007 6.99995C0.320007 6.54661 0.486672 6.15328 0.820007 5.81995C1.15334 5.48661 1.54667 5.31995 2.00001 5.31995ZM2.00001 3.67995C1.54667 3.67995 1.15334 3.51328 0.820007 3.17995C0.486672 2.84661 0.320007 2.45328 0.320007 1.99995C0.320007 1.54661 0.486672 1.15328 0.820007 0.819946C1.15334 0.486611 1.54667 0.319946 2.00001 0.319946C2.45334 0.319946 2.84667 0.486611 3.18001 0.819946C3.51334 1.15328 3.68001 1.54661 3.68001 1.99995C3.68001 2.45328 3.51334 2.84661 3.18001 3.17995C2.84667 3.51328 2.45334 3.67995 2.00001 3.67995Z"
                                fill={appearance ? "white" : "#5F6367"}
                            />
                        </svg>
                    </div>
                </div>
            </div>
        </Frame>
    )
}
