// @framer/steveruizok.education/code/canvas.tsx
// WARNING: this file is auto generated, any changes will be lost
import { createDesignComponent, CanvasStore } from "framer"
const canvas = CanvasStore.shared(); // CANVAS_DATA;
export const Container = createDesignComponent<{parentSize?:{width:number|string,height:number|string},width?:number|string,height?:number|string}>(canvas, "id_x678hlcb8", {}, 375,220);
export const Divider = createDesignComponent<{parentSize?:{width:number|string,height:number|string},width?:number|string,height?:number|string}>(canvas, "id_bqsC9adwp", {}, 375,48);
export const Helpers_Colors = createDesignComponent<{parentSize?:{width:number|string,height:number|string},width?:number|string,height?:number|string}>(canvas, "id_vK6xsDKRe", {}, 375,812);
export const Helpers_Text = createDesignComponent<{parentSize?:{width:number|string,height:number|string},width?:number|string,height?:number|string}>(canvas, "id_T6QtGhQiI", {}, 375,812);
export const Template = createDesignComponent<{parentSize?:{width:number|string,height:number|string},width?:number|string,height?:number|string}>(canvas, "id_XREstntDl", {}, 375,812);
export const Templates_Blank = createDesignComponent<{parentSize?:{width:number|string,height:number|string},width?:number|string,height?:number|string}>(canvas, "id_o00qqOMJZ", {}, 375,812);
export const Templates_Navigation = createDesignComponent<{parentSize?:{width:number|string,height:number|string},width?:number|string,height?:number|string}>(canvas, "id_fNc61tttc", {}, 375,812);
export const Templates_NavigationLarge = createDesignComponent<{parentSize?:{width:number|string,height:number|string},width?:number|string,height?:number|string}>(canvas, "id_OcoXmDPgh", {}, 375,812);
export const Templates_NavigationLarge_TabBar = createDesignComponent<{parentSize?:{width:number|string,height:number|string},width?:number|string,height?:number|string}>(canvas, "id_PeXYyIRny", {}, 375,812);
export const Templates_Navigation_TabBar = createDesignComponent<{parentSize?:{width:number|string,height:number|string},width?:number|string,height?:number|string}>(canvas, "id_hMCzrRmLr", {}, 375,812);

export const colors = Object.freeze({
    /** #0056FE */
    "Primary": "var(--token-f69732ff-3858-4d7e-a918-e92879e5d6bf, rgb(0, 86, 254))",
    /** #03AFE8 */
    "Secondary": "var(--token-c56b99fa-6bea-4887-b827-36fa58a3817f, rgb(3, 175, 232))",
    /** #FFAB57 */
    "Accent": "var(--token-b16a7a8a-e2ab-4afa-a205-534e354689de, rgb(255, 171, 87))",
    /** #FF4366 */
    "Warn": "var(--token-99f5b091-583d-4c54-9ed1-c13fe2b565b5, rgb(255, 67, 102))",
    /** #FFFFFF */
    "Light": "var(--token-c4108478-04be-43e8-b06c-e5db6cfe4ed9, rgb(255, 255, 255))",
    /** #E1E1EF */
    "Neutral": "var(--token-457a7788-6819-4913-8914-5f2a6307a39b, rgb(225, 225, 239))",
    /** #23232F */
    "Dark": "var(--token-655be689-744d-4b94-af37-d052b01faead, rgb(35, 35, 47))",
    /** #131316 */
    "Darker": "var(--token-30f9a7fa-de7c-4aee-a7b8-79cdd74c6221, rgb(19, 19, 22))",
    /** #C9C9D9 */
    "Border": "var(--token-23bce3fd-e7be-45a5-a2c3-53cfc3c53db9, rgb(201, 201, 217))",
    /** rgba(255, 255, 255, 0.68) */
    "Glow": "var(--token-dd5568c7-f380-405b-bbfc-805caca247f3, rgba(255, 255, 255, 0.68))",
    /** rgba(37, 37, 45, 0.09) */
    "Shadow": "var(--token-0b9e3e19-2ca6-4163-8b23-40217b78ddb8, rgba(37, 37, 45, 0.09))",
    /** #BDBDD0 */
    "Active": "var(--token-385efacc-ef57-4c97-b6f0-c789a417631f, rgb(189, 189, 208))",
    /** #FDFDFD */
    "Bg": "var(--token-570013b9-ccd1-4a4b-8a44-e7ded788d7f2, rgb(253, 253, 253))",
})
