// @framer/maxsteitle.resizable-window/code/App.tsx
// import { Data, Override, Animatable, animate } from 'framer';

// const data = Data({
// 	top: Animatable(0),
// 	staticTop: 0,
// 	height: 40,
// 	maxWidth: 500,
// 	width: 100
// });

// export const AddWindowEvents: Override = () => {
// 	return {
// 		onScroll(top) {
// 			data.top = top;
// 			data.staticTop = top;
// 		},
// 		onHeightChange(height) {
// 			data.height = height;
// 		},
// 		onWidthChange(width) {
// 			data.width = width;
// 		}
// 	};
// };

// export const windowHeight: Override = () => {
// 	return {
// 		height: data.height
// 	};
// };

// export const windowWidth: Override = () => {
// 	return {
// 		style: {
// 			width: '100%'
// 		}
// 	};
// };

// export const Sticky: Override = () => {
// 	return {
// 		top: data.top,
// 		willChangeTransform: true
// 	};
// };

// export const StickyWithinRange1: Override = () => {
// 	return {
// 		top: Math.min(Math.max(data.staticTop + 60, 1039), 1360),
// 		willChangeTransform: true
// 	};
// };

// export const StickyWithinRange2: Override = () => {
// 	return {
// 		top: Math.min(Math.max(data.staticTop + 60, 800), 1100),
// 		willChangeTransform: true
// 	};
// };

// export const Parallax1: Override = () => {
// 	return {
// 		top: data.staticTop * 0.3 + 40
// 	};
// };

// export const Parallax2: Override = () => {
// 	return {
// 		top: data.staticTop * 0.48 + 620
// 	};
// };
